---
title: "Assignment3"
author: "Huilin Niu"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: '2'
    df_print: paged
    fig_caption: yes
  html_notebook:
    toc: yes
    toc_depth: '2'
bibliography: A3ref.bib
csl: biomed-central.csl
editor_options: 
  markdown: 
    wrap: 72
params:
  run_gsea: yes
  is_docker: yes
---

# Install and load packages

## Install packages

```{r Install packages, message=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
}

if (!requireNamespace("knitr", quietly = TRUE)){
  install.packages("knitr")}

if (!requireNamespace("RCurl", quietly = TRUE)){
  install.packages("RCurl")}

if (!requireNamespace("ggplot2", quietly = TRUE)){
  install.packages("ggplot2")}

if (!requireNamespace("RCy3", quietly = TRUE)){
  BiocManager::install("RCy3")}

if (!requireNamespace("httr", quietly = TRUE)){
  BiocManager::install("httr")}

if (!requireNamespace("htmltools", quietly = TRUE)){
  install.packages("htmltools")}
```

@r@BiocManager@knitr

## Load packages

```{r Load Packages, message=FALSE}
library("knitr")
library("RCurl")
library("ggplot2")
library("RCy3")
library("httr")
library("htmltools")

```

## Docker Setting

```{r docker setting for running CytoScape}
if(params$is_docker){
  current_base = "host.docker.internal:1234/v1"
  .defaultBaseUrl <- "http://host.docker.internal:1234/v1"
} else{
  current_base = "localhost:1234/v1"
}
```

# Introduction

## Data Exploration

The dataset we used in this assignment here is downloaded from GEO with id [GSE179448](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE179448). The dataset includes RNAseq analysis of human CD4+ regulatory Tregs and Tconvs in COVID-19 and healthy donors isolated from peripheral blood. The dataset was performed on using [GPL18573](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL18573) Illumina NextSeq 500 platform.

## Background

The dataset we used is linked to the published article: "[Profound Treg pertubations correlate with COVID-19 severity](https://www.pnas.org/doi/abs/10.1073/pnas.2111315118?url_ver=Z39.88-2003&rfr_id=ori%3Arid%3Acrossref.org&rfr_dat=cr_pub++0pubmed)". Regulatory T cells (Tregs) are a subset of T cells that play an important role in maintaining immune homeostasis and preventing autoimmunity. Patients with COVID-19 have significant perturbations in Treg populations, which can have important implications for disease pathogenesis and clinical outcomes. This article performed a RNASeq on differnt donor samples at different health stages in regulatory T cells and conventional T cells. @galvan2021

## Data Extraction

There are in total of two supplementary files provided in the dataset called "GSE179448_Gene_count_table.tsv.gz" and "GSE179448_Metadata_GalvanLeon_COVID19_Treg_ULIRNAseq.xlsx".
<br />
The first one contains count data from RNASeq, and the second excel file contains information of the genral information on each sample. The first file is used to extract all count data. The second excel file is used for creating labels and defining groups. Both files are used to group defining and data cleaning processes.

## Data Filtering

There is a total of 86 donor samples in the dataset, where 43 of them are Regulatory T cell samples from donors, and the other 43 samples are conventional T cell samples from donors. There are four groups of donors included: healthy donors(H4), mild outpatients(M3), recovered (R2) and hospitalized to severe(S1). There are 56120 genes in the unprocessed dataset. After filtering out the low count data, I have removed 39466 outliers(count per million(cpm) less than 1). There is 16654 genes left after low count filtering.

## Distribution and Normalization

Comparing the boxplot before and after normalization, it is evident that the log2CPM and median of log2CPM remain relatively stable. However, the median of each sample or replicate becoming more closely aligned.The distribution of the density plot is not perfect as in there is a smaller peak before the main peak. This could be caused by mixing two cell types together in the normalization process.

## Dispersion

The dispersion of the dataset shows variations in the trend line. The variation seen here can come from either biological or technical source in a RNASeq experiment.

## ID Mapping

There are a total of 2154 expression data not mapped. After this data cleaning and filtering process, we have 14500 normalized gene expression left excluding the ones are not being mapped 2154 expression. They make up the 16654 genes after removing low counts. In this assignment, we will continue to perform differntial gene analysis.

## Differential gene expression

In the differential gene expression analysis, I used the limma package to preform the scoring of the dataset. I have tried both the simple model(only consider cell type) and complex model(consider both the cell type and the donor samples) for modelling the data set. From analysis the data using MDS plot. The complex model makes more sense because donor sample seems to be a contributor when looking at the MDS plot as samples from the same donor group seems to cluster together. I used Benjamini-Hochberg (fdr) method for multiple hypothesis testing which will control the false discovery rate of our results. When we use the threshold of 0.001, there are 4339 deferentially expressed genes, and 3730 of them pass the correction.

```{r render annotated heatmap, fig.cap="Figure 1: Heatmap for differential expressed genes. The expression level are shown in a color scale from blue to red(from low to high). The differntial expressed genes are analyzed using limma, genes selected here used the threshold of p < 0.001. The red box marks the profound regulatory t cells pertubations in donors, and the yellow box marks the profound conventional t cells pertubations in donors."}
knitr::include_graphics(file.path(getwd(), "figures",
              "t_heatmap_for_differntial_expressed_genes_annotated.png"))
```

As seen in the figure above, the samples cluster nicely by cell type. If we look at the data closely, it also cluster by donor group. The left side of clustered samples represent regulatory T cell samples from donor samples, whereas the right side of the clustered samples represent conventional T cell samples from donor samples. These two clusters have very different gene expressions. The top halves of the genes are highly expressed in regulatory T cells, but the bottom halves of the genes are highly expressed in conventional T cells. If we take a look at the left and the right halves of the figure separately, we can see differential cell expressions depending on the condition of the donor samples. The red box indicates the differential expressed genes in regulatory T cells. If we take a look at the sample labels at the bottom, the samples with high expressions mostly consists of S1(hospitalized to severe) and R3(recovered) samples. This supports the published articles discovery that there are profound differences in gene expression depends on the donor conditions. We will be focused on this group in the analysis going forward. The yellow box indicates differntial expressed genes in conventional T cells. If we take a look at the sample labels at the bottom, the samples with high expressions mostly consists of S1(hospitalized to severe) and R3(recovered) samples. This shows that there is also differences in gene expression in conventional T cells depends on the donor conditions. However, there are different groups of genes being expressed differently.

<br /> 
After performing the differential gene set analysis, we generated three files containing all ranked genes, upregulated genes(fold change greater than 0), and downregulated genes(fold change less than 0). 

## Threshold Over-representation analysis

I chose g:Profiler to analyze the data. I used <GO:BP> in Gene Ontology databse, KEGG, Reactome and WikiPathway to annotate my data. The the databases were the following version: (1)hsapien(Human) -version: GRCh38.p13; (2) <GO:BP>: annotations: BioMart (releases/2022-12-04); (3) KEGG:KEGG FTP Release 2022-12-26; (4)REAC: annotations: BioMart (releases/2022-12-28); (5)WikiPathways: 20221210 
<br /> 
Threshold is set at 0.05 for g:Profiler analysis, the terms are set to from 8-300. 
<br /> 
Ranked_genes:There are 1416 <GO:BP> terms found, 2 KEGG terms, 5 REAC terms, and 5 WP terms found in ranked gene list. Let's take a closer look by looking at the detailed values. 
<br /> 
Unregulated_genes: There are 473 <GO:BP> terms found, 2 KEGG terms, 136 REAC terms, and 4 WP terms found in upregulated genes list. <br /> Downregulated_genes:There are 1842 <GO:BP> terms found, 9 KEGG terms, 14 REAC terms, and 6 WP terms found in downregulated genes list. 
<br /> 
The over-representation results support the conclusion and the mechanism discussed in the paper. I have identified multiple pathways that are related to the immune system and cell proliferation pathway. The downregulated portion of genes have enriched terms like immune response, which supports the paper's discovery of very severe T cell expression differences depending on the condition of patients. The upregulated portion of genes have enriched terms like cell proliferation which suggest rapid cell division which can be explained by the virus replication processes.

# Non-thresholded Gene set Enrichment Analysis

## Procedure for running GSEA inside java

### Download the pathway file

The gmt file we used in this analysis will be downloaded from the Bader Lab database. 
```{r download data, message=FALSE, warning=FALSE}
# Download data from Bader Lab Gene Set
# This url is directed to the most current release of the human subdirectory in the bader lab file
gmt_file_url = "http://download.baderlab.org/EM_Genesets/current_release/Human/symbol/"

# list all the files on the server
allfiles = RCurl::getURL(gmt_file_url)

# create a text connection for input
input = textConnection(allfiles)
contents = readLines(input)

# close the connection
close(input)

# get the gmt that has all the pathways and does not include terms inferred from electronic annotations(IEA)
# start with gmt file that has pathways only, return all match data from the list of the files
matches = gregexpr(pattern = "(?<=<a href=\")(.*.GOBP_AllPathways_no_GO_iea.*.)(.gmt)(?=\">)",
              text = contents,
              perl = TRUE)

# extract the name of the file we wanted
gmt_file = unlist(regmatches(x = contents, 
                             m = matches))

# The full file path of the gmt file to be downloaded
downloaded_gmt_file <- file.path(getwd(),"data",gmt_file)

# check if the gmt file is already in the directory, if it is, do not download again.
if(!file.exists(downloaded_gmt_file)){
  # download the file if it has not been downlaoded
  download.file(url = paste0(gmt_file_url,gmt_file), 
                destfile = downloaded_gmt_file)
}

```


### Configurate Parameters

Before we run GSEA from R, we first need to set a couple parameters to make sure it runs smoothly.

```{r initialize parameters}
# Set parameters for initialize GSEA
#directory of the gsea in docker
gsea_jar <- "/home/rstudio/GSEA_4.3.2/gsea-cli.sh"
java_version <- 11

# set to false to speed up knitting, and false after have the intial results
run_gsea = params$run_gsea

# Leave blank for R to find its directory
gsea_directory = ''
rnk_file <- file.path(getwd(), "data", "ranked_gene_list_all.rnk")

analysis_name <- "Treg VS Tconv"
is_docker <- params$is_docker
```


### Run GSEA from R
After setting the parameter, we will continue to run gsea by calling the java program using system command.
```{r Run GSEA, warning=FALSE, message=FALSE}
# The file path of the gsea_output file
result_file <- file.path(getwd(), "gsea_output.txt")

# Check if GSEA is already here, if it is, do not run again.
if(!file.exists(result_file)){
  # If the file already exist, do not run the gsea analysis again
  if(run_gsea && java_version == "11"){
  # Set command for running gsea using the java the program
  # -rnk set: path to rnk file
  # -scoring_scheme: set scoring scheme to weighted
  # -collapse false: do not need to collapse because the data is already cleaned
  # -nperm: 2000 Set the number of permutation to 2000
  # -set_max: Set the max term as 200 to get more meaningful results
  # -out: set output path
  gsea_command <- paste("",
                   gsea_jar,  
                   "GSEAPreRanked -gmx", 
                   downloaded_gmt_file, 
                   "-rnk",
                   rnk_file, 
                   "-collapse false -nperm 2000 -scoring_scheme weighted -rpt_label ",
                   analysis_name,
                   "  -plot_top_x 20 -rnd_seed 12345 -set_max 200 -set_min 15 -zip_report false -out",
                   file.path(getwd(), "results"),
                   " > gsea_output.txt",
                   sep=" ")
  system(gsea_command)
  }
}
```

### Get the output results from GSEA

After running the GSEA, now we will need to find the gsea output directory. Because we will make sure GSEA is only run once in the R notebook only once. Therefore, we will only need to locate to that GSEA file.
```{r find GSEA directory}
if(gsea_directory == ""){
  # Find the exact working directory of gsea
  gsea_directory <- list.files(path = file.path(getwd(), "results"), 
                                 pattern = "\\.GseaPreranked")
  # Because we have set the parameters to only run once, we do not need to find the newest file anymore.
}
```

### Analyze the results from the analysis

```{r gsea results}
# Find both the up-regulated and down-regulated terms in these two files.

new_gsea_directory <- file.path(getwd(),"results",gsea_directory)
# html File name for negatively enriched terms
neg_reports <- list.files(path = new_gsea_directory,
                      pattern = "^gsea_report_for_na_neg")[1]
# TSV File name for negatively enriched terms
neg_tsv <- list.files(path = new_gsea_directory,
                      pattern = "^gsea_report_for_na_neg")[2]

# html File name for positively enriched terms
pos_reports <- list.files(path = new_gsea_directory,
                      pattern = "^gsea_report_for_na_pos")[1]

# TSV File name for positively enriched terms
pos_tsv <- list.files(path = new_gsea_directory,
                      pattern = "^gsea_report_for_na_pos")[2]

# File name for index page
index_page <- list.files(path = new_gsea_directory,
                      pattern = "^index")
# File path for index page
index_page <- file.path(new_gsea_directory, index_page)
```
We will continue the analysis of this section later in question 2. Before take a look at the results, I also tried another way to run gsea through the Java software. Of course, they will generate the same results.
## Procedure for running in the java software

The following figures shows the procedures and parameters for running
the GSEA analysis. In this analysis, we used the most current release of
the GSEA software version: v4.3.2[build: 13]. @gsea

```{r GSEA procedure 1, fig.cap="Figure 2: Step 1-2 of running GSEA analysis"}
knitr::include_graphics(file.path(getwd(), "figures",
              "GSEA_procedure_1.png"))
```

**Step 1:** First, load the gmt file containing all pathways, and the
geneset file "ranked_gene_list_all.rnk" by using Method 1(marked by the
red box, and labelled Step 1). <br /> **Step 2:** Second, click "Run
GSEAPreranked" button to set the parameters to analyze the preranked
file.

```{r GSEA procedure 2, fig.cap="Figure 3: Step 3-5 of running GSEA analysis."}
knitr::include_graphics(file.path(getwd(), "figures",
              "GSEA_procedure_2.png"))
```

**Step 3:** In the required field section, the gene set database we used is downloaded from the [Bader Lab Gene Set](https://download.baderlab.org/EM_Genesets/current_release/Human/symbol/). The number of permutation is increased to 2000 for a more stringent analysis. The ranked list field is loaded with the  "ranked_gene_list_all.rnk". The Collapse/Remap to gene symbols was selected as no_collapse as we have performed data cleaning, filtering and normalization for the dataset in previous assignments. 
<br /> 
**Step 4:** In the basic fields section, the Analysis name is filled with "Treg VS Tconv", the enrichment statistics is set is weighted. The "max size: exclude larger sets" field is decreased to 200 to exclude larger sets and get more meaningful results.
<br /> 
**Step 5:** Click run to run the GSEA analysis. 
<br /> 


## Discussion Questions
**Question 1: What method did you use? What genesets did you use?** 

Here, we used GSEA software for the non-thresholded gene set enrichment analysis. The genesets file "ranked_gene_set_all.rnk" was used for the analysis. This file contains all differntially expressed genes found in the differential analysis in Assignment 2 with no threshold for p value or fold change. 

```{r ranked_gene_list file, fig.cap="Figure 4: Ranked Genelist for "}
ranked_file <- read.table(file.path(getwd(), "data",
                                    "ranked_gene_list_all.rnk"))
colnames(ranked_file) <- c("GeneSymbol", "FoldChange")
knitr::kable(ranked_file[1:5,])
```


The ranked_gene_list_all.rnk file contains two columns, the first column is the gene symbol, and the second column is the fold change. The file is sorted by the fold change column in a decreasing order.
```{r}
#Number of downregulated genes
length(which(ranked_file$FoldChange < 0))

#Number of upregulated genes
length(which(ranked_file$FoldChange > 0))

# Total number of genes
nrow(ranked_file)
```

We have included a total number of 14094 genes in the GSEA analysis, in which there are 6867 genes are downregulated and 7227 genes are upregulated. 

In this analysis, we used the most current release of the GSEA software version: v4.3.2[build: 13] @gsea

<br /> 

**Question 2: Summarize your enrichment results.** 
The following shows a summary of the GSEA analysis we have run.
```{r gsea index}
htmltools::includeHTML(index_page)
```

The dataset has 14094 features (genes). By setting the term size to between 15 and 200, we filtered out 14300 / 19708 gene sets. 
<br />
For the positively enriched terms set, there are 2332 / 5408 gene sets are upregulated, 3 gene sets are significant at FDR < 25%, 39 gene sets are significantly enriched at nominal pvalue < 1%, 183 gene sets are significantly enriched at nominal pvalue < 5%.
<br />
For the negatively enriched terms set, there are 3076 / 5408 gene sets are upregulated, 88 gene sets are significantly enriched at FDR < 25%, 93 gene sets are significantly enriched at nominal pvalue < 1%, 248 gene sets are significantly enriched at nominal pvalue < 5%.

**The following shows a summary of the positively enriched terms set.**
```{r gsea pos results}
htmltools::includeHTML(file.path(new_gsea_directory, pos_reports))
```

Here, from the above results. The fourth row shows "SARS-COV-2 ACTIVATES MODULATES INNATE AND ADAPTIVE IMMUNE RESPONSES" and the 13th row "REGULATION OF G0 TO G1 TRANSITION". These two results are particularly interesting because it seems to be able to explain by biology behind the data. We will explain this further in the interpretation question. Overall, the terms seems to points to pathways related to cell junction and cell proliferation.
**The following shows a summary of the negatively enriched terms set.**
```{r gsea neg results}
htmltools::includeHTML(file.path(new_gsea_directory, neg_reports))
```
In the negatively enriched term sets, I do not find any terms of particular interest. The terms are all generally related to the cell respiratory system and mitochondria. It made sense because the experiment is done on COVID and healthy donors, and COVID-19 is a respiratory disease which could affect the cell respiratory and mitochondria function to generate ATP.

<br />

**Question 3: How do these results compare to the results from the thresholded analysis in Assignment #2. Compare qualitatively. Is this a straight forward comparison? Why or why not?**

It seems to be a relatively straight forward comparison when comparing both results. <br />
Both thresholded and non-thresholded analysis do seem to have some similarity in terms of the results they have generated. For exmaple. both analysis get results related to cell respiratory pathway, cell proliferation and immune system. However, there are many differences regarding the specificity of the terms generated and which terms are the most significant. 
<br />
Overall, I think the non-thresholded analysis gave a better result than the non-thresholded one we done on g:Profiler, but both results supports the conclusion of the paper.


# Visualize Gene set Enrichment Analysis in Cytoscape

## Launch Cytoscape

First, to run Cytoscape from R, we need to first connect R to Cytoscape using "RCy3" package.

```{r launch cytoscape}
RCy3::cytoscapePing(base.url = current_base)
RCy3::cytoscapeVersionInfo(base.url = current_base)
```

The above result showed that we have successfully connected to Cytoscape.
## Create an Enrichment map
```{r}
# Upload EM file function in the local path.
upload_em_file <- function(local_path) {
    bname <- basename(local_path)
    postfile_to_server <- POST(
      # url to docker
      url = paste('http://host.docker.internal:1234/enrichmentmap/textfileupload?fileName=', bname, sep=""),
      # no additional cofig info
      config = list(),
      # body being the upload_file
      body = list(file = upload_file(local_path)),
      # here we used multipart(multipart/form-data)
      encode = "multipart",
      handle = NULL
    )
    content(postfile_to_server,"parsed")$path
}
```


```{r set parameters for enrichment map}
pvalue_em <- "0.005"
qvalue_em <- "0.5"

similarity_em <- "0.375"
similarity_metric_em = "COMBINED"

network_name <- analysis_name

edb_results_path <- file.path(new_gsea_directory,"edb", "results.edb")

#if using docker we need to replace all the the paths to the host path
if(is_docker) {
  # "upload" the files to the host machine and replace each path with the host machine path
  gmt_gsea_uploaded <- upload_em_file(downloaded_gmt_file)
  gsea_ranks_uploaded <- upload_em_file(rnk_file)
  gsea_results_uploaded <- upload_em_file(edb_results_path)
}
```



```{r create enrichment map}
current_network_name <- paste(network_name,pvalue_em,qvalue_em,sep="_")
em_command = paste('enrichmentmap build analysisType="gsea" gmtFile=',
                   gmt_gsea_uploaded,
                   'pvalue=',
                   pvalue_em, 
                   'qvalue=',
                   qvalue_em,
                   'similaritycutoff=',
                   similarity_em,
                   'coefficients=',
                   similarity_metric_em,
                   'ranksDataset1=', 
                   gsea_ranks_uploaded,
                   'enrichmentsDataset1=',
                   gsea_results_uploaded, 
                   'filterByExpressions=false',
                   'gmtFile=',gmt_gsea_uploaded,
                   sep=" ")
#enrichment map command will return the suid of newly created network.
response <- commandsGET(em_command,base.url = current_base)
current_network_suid <- 0
#enrichment map command will return the suid of newly created network unless it Failed.  
#If it failed it will contain the word failed
if(grepl(pattern="Failed", response)){
  paste(response)
} else {
  current_network_suid <- response
}
#check to see if the network name is unique
current_names <- getNetworkList(base.url = current_base)
if(current_network_name %in% current_names){
  #if the name already exists in the network names then put the SUID in front
  # of the name (this does not work if you put the suid at the end of the name)
  current_network_name <- paste(current_network_suid,current_network_name,  sep="_")
}
response <- renameNetwork(title=current_network_name, 
                       network = as.numeric(current_network_suid),base.url = current_base)
```

```{r get screenshot of the initial network}
# When we are not running the rmd in the docker, we can generate a file in the figures directory.
if(!params$is_docker){
  output_network_file <- file.path(getwd(),"figures","initial_screenshot_network.png")
  output_network_file_current <- output_network_file
  fitContent()
  if(file.exists(output_network_file)){
  #cytoscape hangs waiting for user response if file already exists.  Remove it first
  response <- file.remove(output_network_file)
  } 
  response <- exportImage(output_network_file, type = "png",base.url = current_base)
}

```

**Question 1: Create an enrichment map - how many nodes and how many edges in the resulting map? What thresholds were used to create this map? Make sure to record all thresholds. Include a screenshot of your network prior to manual layout.**

This step is done by a couple steps of optimization. I have used 0.5 FDR q-value cutoff, p-value cutoff 0.01, Similarity threshold 0.375, similarity scoring metric: combined. 

<br />

I have tried a few fdr q-value cutoff, when fdr is 0.01, there is no upregulated pathways because all upregulated pathways does not pass correction. I further increase the fdr to 0.25, and there are only 3 gene sets. Finally, I set the fdr threshold at 0.5 which I think gave more results about the upregulated pathways. 

<br />

There are 100 nodes and 1822 edges in the network.

```{r rendeinitial screenshot, fig.cap="Figure 2: Initial screenshot of the network before manual layout."}
knitr::include_graphics(file.path(getwd(), "figures",
              "intial_screenshot.png"))
```


**Question 2: Annotate your network - what parameters did you use to annotate the network. If you are using the default parameters make sure to list them as well.**

```{r render annotation parameter1, fig.cap="Figure 3: Parameter settings in autoannotation in Cytoscape"}
knitr::include_graphics(file.path(getwd(), "figures",
              "autoannotation_parameters.png"))
```
I first tired to annotate the network by using autoannotation App in Cytoscape, the above image showed the parameters I used. Most of the parameters I used are default. The only differnece is that I set Edge weight column to similarity_coefficient. Because the clusters seemed too general, therefore, I mannually annoteted the nodes.

```{r render annotated 1, fig.cap="Figure 4:Annotated Network Treg VS Tconv Sample Expression with Label in Cytoscape"}
knitr::include_graphics(file.path(getwd(), "figures",
              "with_label.png"))
```


**Question 3: Make a publication ready figure - include this figure with proper legends in your notebook.**

```{r render annotated 2, fig.cap="Figure 5:Annotated Network Treg VS Tconv Sample Expression without Label in Cytoscape"}
knitr::include_graphics(file.path(getwd(), "figures",
              "without_label.png"))
```

**Question 4: Collapse your network to a theme network. What are the major themes present in this analysis? Do they fit with the model? Are there any novel pathways or themes?**

```{r render annotated 3, fig.cap="Figure 6:Collapsed Annotated Network Treg VS Tconv Sample Expression without Label in Cytoscape"}
knitr::include_graphics(file.path(getwd(), "figures",
              "collapsed_theme.png"))
```

**Post Analysis**
```{r render Post analysis 1, fig.cap="Figure 7:Parameters for Post Analysis. Here, only immune system and cardiovascular system is chosen."}
knitr::include_graphics(file.path(getwd(), "figures",
              "hp_sig.png"))
```


```{r render Post analysis 2, fig.cap="Figure 7:Post Analysis Network Treg VS Tconv Sample Expression without Label in Cytoscape."}
knitr::include_graphics(file.path(getwd(), "figures",
              "hp_phenotype.png"))
```


# Intepretation and detailed view of results

**Question 1: Do the enrichment results support conclusions or mechanism discussed in the original paper? How do these results differ from the results you got from Assignment #2 thresholded methods?**

Yes, the enrichment results supports the conclusion in the paper, which is profound regulatory T cells and conventional T cells pertubations among samples in different coditions. However, I found the results from GSEA analysis are better from those in g:Profiler. They are able to provide more detials regarding more specific pathways, and matched to more specific terms that is related to the paper "SARS-Cov2".

**Question 2: Can you find evidence, i.e. publications, to support some of the results that you see. How does this evidence support your result?**

Yes, the results are suppotive of the article. Multiple pathways related to the immune system are found. There are multiple publications that also find evidence that immune response are downregulated in COVID patients which supports my findings here. However, there is no evidence showing that there will be upregulation of cell proliferation in COVID patients. This could be explained by how the virus can use cell cycle arrest to promote replication of viral factors, for example the enriched terms on regulation of G0/G1 transition in our gsea analysis.@immune

#Reference
